File: packages/cli/src/instruction-set-loader.ts
import { existsSync, realpathSync, readFileSync, readdirSync, statSync } from 'node:fs'
import { resolve, sep } from 'node:path'

export class InstructionSetLoaderError extends Error {
  constructor(readonly code: string) { super(code); this.name = 'InstructionSetLoaderError' }
}

const LEGACY_LOADER_RETIRED = 'instruction_set_legacy_loader_retired: this tree was synced by the paved path; see docs/cli.md#migrating-the-legacy-sync-layout'

interface Material { files: Record<string, string>; pinnedComponents?: Record<string, unknown> }
interface Manifest { manifestVersion: number; projectId?: string; instructionSetId?: string; name: string; slug?: string; shape: string[]; default: Material; profiles: Array<{ modelConfigIdentity: string; production: Material | null }>; filteredTo?: string[] }

function inside(root: string, candidate: string) { return candidate === root || candidate.startsWith(`${root}${sep}`) }

function manifestMatchesReference(candidate: string, reference: string): boolean {
  try {
    const manifest = JSON.parse(
      readFileSync(resolve(candidate, 'manifest.json'), 'utf8')
    ) as Pick<Manifest, 'name' | 'slug'>
    return manifest.name === reference || manifest.slug === reference
  } catch {
    return false
  }
}

function pavedTreeMatchesDisplayName(directoryRoot: string, reference: string): boolean {
  const setsRoots = [
    resolve(directoryRoot, 'orizu', 'instruction-sets'),
    resolve(directoryRoot, 'instruction-sets'),
  ]
  for (const setsRoot of setsRoots) {
    try {
      for (const setEntry of readdirSync(setsRoot, { withFileTypes: true })) {
        if (!setEntry.isDirectory()) continue
        const setRoot = realpathSync(resolve(setsRoot, setEntry.name))
        if (!inside(directoryRoot, setRoot)) continue
        for (const profileEntry of readdirSync(setRoot, { withFileTypes: true })) {
          if (!profileEntry.isDirectory()) continue
          const profileRoot = realpathSync(resolve(setRoot, profileEntry.name))
          if (!inside(directoryRoot, profileRoot)) continue
          for (const versionEntry of readdirSync(profileRoot, { withFileTypes: true })) {
            if (!versionEntry.isDirectory() || !/^v[1-9][0-9]*$/u.test(versionEntry.name)) continue
            try {
              const manifest = JSON.parse(readFileSync(resolve(profileRoot, versionEntry.name, 'manifest.json'), 'utf8')) as {
                instructionSetName?: unknown
                instructionSetSlug?: unknown
              }
              if (manifest.instructionSetName === reference || manifest.instructionSetSlug === reference) return true
            } catch {
              // Inspect another paved Version.
            }
          }
        }
      }
    } catch {
      // Inspect the other paved root.
    }
  }
  return false
}

function resolveSetRoot(directory: string, reference: string): string {
  let directoryRoot: string
  try { directoryRoot = realpathSync(directory) } catch { throw new InstructionSetLoaderError('instruction_set_not_synced') }
  const pavedCandidates = [
    resolve(directoryRoot, 'orizu', 'instruction-sets', reference),
    resolve(directoryRoot, 'instruction-sets', reference),
  ]
  if (pavedCandidates.some(candidate => {
    try { return inside(directoryRoot, realpathSync(candidate)) && statSync(candidate).isDirectory() } catch { return false }
  }) || (!/^[a-z0-9]+(?:-[a-z0-9]+)*$/u.test(reference)
    && pavedTreeMatchesDisplayName(directoryRoot, reference))) {
    throw new InstructionSetLoaderError(LEGACY_LOADER_RETIRED)
  }
  const direct = resolve(directoryRoot, reference)
  try {
    if (inside(directoryRoot, direct) && existsSync(direct) && statSync(direct).isDirectory()) {
      const realDirect = realpathSync(direct)
      if (inside(directoryRoot, realDirect)) {
        try {
          const manifest = JSON.parse(
            readFileSync(resolve(realDirect, 'manifest.json'), 'utf8')
          ) as Pick<Manifest, 'name' | 'slug'>
          if (manifest.name === reference || manifest.slug === reference) {
            return realDirect
          }
        } catch {
          // Preserve the exact-path loader's specific missing/invalid manifest error.
          return realDirect
        }
      }
    }
  } catch {
    // A stale or unreadable direct-path candidate must not shadow a valid set.
  }
  const matches = readdirSync(directoryRoot, { withFileTypes: true }).flatMap(entry => {
    try {
      if (entry.name.startsWith('.') || !entry.isDirectory()) return []
      const candidate = realpathSync(resolve(directoryRoot, entry.name))
      return inside(directoryRoot, candidate) && manifestMatchesReference(candidate, reference)
        ? [candidate]
        : []
    } catch {
      return []
    }
  })
  if (matches.length !== 1) throw new InstructionSetLoaderError(
    matches.length > 1 ? 'instruction_set_reference_ambiguous' : 'instruction_set_not_synced'
  )
  return matches[0]!
}

export function loadInstructionSet(directory: string, name: string, modelConfigIdentity: string): Record<string, string> {
  const setRoot = resolveSetRoot(directory, name)
  const manifestPath = resolve(setRoot, 'manifest.json')
  if (!existsSync(manifestPath)) throw new InstructionSetLoaderError('instruction_set_manifest_missing')
  let manifest: Manifest
  try { manifest = JSON.parse(readFileSync(manifestPath, 'utf8')) as Manifest } catch { throw new InstructionSetLoaderError('instruction_set_manifest_missing') }
  if (manifest.manifestVersion !== 1) throw new InstructionSetLoaderError('instruction_set_manifest_version_unsupported')
  if (manifest.filteredTo && !manifest.filteredTo.includes(modelConfigIdentity)) throw new InstructionSetLoaderError('instruction_set_profile_not_synced')
  const profile = manifest.profiles.find(profile => profile.modelConfigIdentity === modelConfigIdentity)
  if (!profile) throw new InstructionSetLoaderError('instruction_set_profile_not_found')
  if (!profile.production) throw new InstructionSetLoaderError('instruction_set_profile_not_promoted')
  const selected = profile.production
  const result: Record<string, string> = Object.create(null)
  for (const key of manifest.shape) {
    if (selected.pinnedComponents?.[key]) throw new InstructionSetLoaderError('instruction_set_component_unavailable')
    const relativePath = selected.files?.[key]
    if (!relativePath) throw new InstructionSetLoaderError('instruction_set_profile_key_missing')
    const unresolvedPath = resolve(setRoot, relativePath)
    if (!inside(setRoot, unresolvedPath)) throw new InstructionSetLoaderError('instruction_set_path_unsafe')
    let path: string
    try { path = realpathSync(unresolvedPath) } catch { throw new InstructionSetLoaderError('instruction_set_component_unreadable') }
    if (!inside(setRoot, path)) throw new InstructionSetLoaderError('instruction_set_path_unsafe')
    try { result[key] = readFileSync(path, 'utf8') } catch { throw new InstructionSetLoaderError('instruction_set_component_unreadable') }
  }
  return result
}

File: packages/orizu-gepa/src/orizu_gepa_connector/instruction_set_loader.py
"""Offline reader for the ALI-1532 synced instruction-set layout."""

import json
import re
from pathlib import Path


LEGACY_LOADER_RETIRED = "instruction_set_legacy_loader_retired: this tree was synced by the paved path; see docs/cli.md#migrating-the-legacy-sync-layout"


class InstructionSetLoaderError(Exception):
    def __init__(self, code: str):
        self.code = code
        super().__init__(code)


def _inside(root: Path, candidate: Path) -> bool:
    try:
        candidate.relative_to(root)
        return True
    except ValueError:
        return False


def _paved_tree_matches_display_name(directory_root: Path, reference: str) -> bool:
    sets_root = directory_root / "orizu" / "instruction-sets"
    try:
        for manifest_path in sets_root.glob("*/*/v*/manifest.json"):
            if not _inside(directory_root, manifest_path.resolve()):
                continue
            try:
                manifest = json.loads(manifest_path.read_text())
            except (OSError, json.JSONDecodeError):
                continue
            if manifest.get("instructionSetName") == reference or manifest.get("instructionSetSlug") == reference:
                return True
    except OSError:
        return False
    return False


def _resolve_set_root(directory: str, reference: str) -> Path:
    directory_root = Path(directory).resolve()
    if not directory_root.is_dir():
        raise InstructionSetLoaderError("instruction_set_not_synced")
    paved_candidates = (
        directory_root / "orizu" / "instruction-sets" / reference,
        directory_root / "instruction-sets" / reference,
    )
    if (
        any(candidate.is_dir() and _inside(directory_root, candidate.resolve()) for candidate in paved_candidates)
        or (not re.fullmatch(r"[a-z0-9]+(?:-[a-z0-9]+)*", reference)
            and _paved_tree_matches_display_name(directory_root, reference))
    ):
        raise InstructionSetLoaderError(LEGACY_LOADER_RETIRED)
    direct = (directory_root / reference).resolve()
    if _inside(directory_root, direct) and direct.is_dir():
        try:
            direct_manifest = json.loads((direct / "manifest.json").read_text())
        except (OSError, json.JSONDecodeError):
            # Preserve the exact-path loader's specific missing/invalid manifest error.
            return direct
        if direct_manifest.get("name") == reference or direct_manifest.get("slug") == reference:
            return direct
    matches: list[Path] = []
    for child in directory_root.iterdir():
        candidate = child.resolve()
        if child.name.startswith(".") or not child.is_dir() or not _inside(directory_root, candidate):
            continue
        try:
            manifest = json.loads((candidate / "manifest.json").read_text())
        except (OSError, json.JSONDecodeError):
            continue
        if manifest.get("name") == reference or manifest.get("slug") == reference:
            matches.append(candidate)
    if len(matches) != 1:
        code = "instruction_set_reference_ambiguous" if len(matches) > 1 else "instruction_set_not_synced"
        raise InstructionSetLoaderError(code)
    return matches[0]


def load_instruction_set(directory: str, name: str, model_config_identity: str) -> dict[str, str]:
    root = _resolve_set_root(directory, name)
    manifest_path = root / "manifest.json"
    if not manifest_path.is_file():
        raise InstructionSetLoaderError("instruction_set_manifest_missing")
    try:
        manifest = json.loads(manifest_path.read_text())
    except (OSError, json.JSONDecodeError) as error:
        raise InstructionSetLoaderError("instruction_set_manifest_missing") from error
    if manifest.get("manifestVersion") != 1:
        raise InstructionSetLoaderError("instruction_set_manifest_version_unsupported")
    if manifest.get("filteredTo") and model_config_identity not in manifest["filteredTo"]:
        raise InstructionSetLoaderError("instruction_set_profile_not_synced")
    profile = next((profile for profile in manifest.get("profiles", []) if profile.get("modelConfigIdentity") == model_config_identity), None)
    if profile is None:
        raise InstructionSetLoaderError("instruction_set_profile_not_found")
    production = profile.get("production")
    if production is None:
        raise InstructionSetLoaderError("instruction_set_profile_not_promoted")
    material = production
    values: dict[str, str] = {}
    for key in manifest.get("shape", []):
        if key in material.get("pinnedComponents", {}):
            raise InstructionSetLoaderError("instruction_set_component_unavailable")
        relative = material.get("files", {}).get(key)
        if not relative:
            raise InstructionSetLoaderError("instruction_set_profile_key_missing")
        path = (root / relative).resolve()
        if not _inside(root, path):
            raise InstructionSetLoaderError("instruction_set_path_unsafe")
        try:
            values[key] = path.read_text()
        except OSError as error:
            raise InstructionSetLoaderError("instruction_set_component_unreadable") from error
    return values
